unittests/Expr/ExprTest.cpp

```diff
@@ -113,4 +113,120 @@ TEST(ExprTest, ExtractConcat) {
   EXPECT_EQ(Expr::Extract, concat2->getKid(1)->getKind());
 }
 
+TEST(ExprTest, ReadExprFoldingBasic) {
+  unsigned size = 10;
+
+  // Constant array
+  std::vector<ref<ConstantExpr> > Contents(size);
+  for (unsigned i = 0; i < size; ++i)
+    Contents[i] = ConstantExpr::create(i + 1, Expr::Int8);
+  ArrayCache ac;
+  const Array *array =
+      ac.CreateArray("arr", size, &Contents[0], &Contents[0] + size);
+
+  // Basic constant folding rule
+  // Constant index (0)
+  ref<Expr> index = ConstantExpr::create(0, Expr::Int32);
+  UpdateList ul(array, 0);
+  ref<Expr> read = ReadExpr::create(ul, index);
+  // Read - should be constant folded to Contents[0]
+  // Check that constant folding worked
+  EXPECT_EQ(Expr::Constant, read.get()->getKind());
+  ASSERT_TRUE(EqExpr::create(Contents[0], read).get()->isTrue());
+}
+
+TEST(ExprTest, ReadExprFoldingIndexOutOfBound) {
+  unsigned size = 10;
+
+  // Constant array
+  std::vector<ref<ConstantExpr> > Contents(size);
+  for (unsigned i = 0; i < size; ++i)
+    Contents[i] = ConstantExpr::create(i + 1, Expr::Int8);
+  ArrayCache ac;
+  const Array *array =
+      ac.CreateArray("arr", size, &Contents[0], &Contents[0] + size);
+
+  // Constant folding rule w index-out-of-bound
+  // Constant index (0)
+  ref<Expr> index = ConstantExpr::create(128, Expr::Int32);
+  UpdateList ul(array, 0);
+  ref<Expr> read = ReadExpr::create(ul, index);
+  // Read - should not be constant folded
+  // Check that constant folding was not applied
+  EXPECT_EQ(Expr::Read, read.get()->getKind());
+}
+
+TEST(ExprTest, ReadExprFoldingConstantUpdate) {
+  unsigned size = 10;
+
+  // Constant array
+  std::vector<ref<ConstantExpr> > Contents(size);
+  for (unsigned i = 0; i < size; ++i)
+    Contents[i] = ConstantExpr::create(i + 1, Expr::Int8);
+  ArrayCache ac;
+  const Array *array =
+      ac.CreateArray("arr", size, &Contents[0], &Contents[0] + size);
+
+  // Constant folding rule w constant update
+  // Constant index (0)
+  ref<Expr> index = ConstantExpr::create(0, Expr::Int32);
+  UpdateList ul(array, 0);
+  ref<Expr> updateValue = ConstantExpr::create(32, Expr::Int8);
+  ul.extend(index, updateValue);
+  ref<Expr> read = ReadExpr::create(ul, index);
+  // Read - should be constant folded to 32
+  // Check that constant folding worked
+  EXPECT_EQ(Expr::Constant, read.get()->getKind());
+  ASSERT_TRUE(EqExpr::create(updateValue, read).get()->isTrue());
```

Comment:
Ok, now I remember why I chose to use EqExpr: you need to cast the read to use those methods and in my opinion that decreases readability.
Do you have any suggestion?